### PR TITLE
Change `Read-Host -MaskInput` to use existing SecureString path, but return as plain text

### LINKED ISSUE #13011: `Read-Host -MaskInput -Prompt 'foo'` doesn't mask the input
https://github.com/PowerShell/PowerShell/pull/10908 added a `-MaskInput` switch to `Read-Host` to allow reading sensitive input from the user.  However, when used with `-Prompt`, the `-MaskInput` switch is no longer in effect thus showing what the user is typing.

## Steps to reproduce

```powershell
read-host -MaskInput -Prompt "Tell me your secret"
```

## Expected behavior

```none
Tell me your secret: ****
```

## Actual behavior

```none
Tell me your secret: test
```

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.1.0-preview.3
PSEdition                      Core
GitCommitId                    7.1.0-preview.3-90-g3d003cb227ebe165d450f558ea48876473050086
OS                             Darwin 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

https://github.com/PowerShell/PowerShell/pull/10908 introduced a new `-MaskInput` switch to enable hiding the input while still returning a string.  It did this by introducing a new method `ReadLineMaskedAsString()`, however, this was not necessary and would not be supported by remoting as it required hosts to implement it.  It also did not support use of `-MaskInput` with `-Prompt`.

The change here is to revert most of the changes in that PR to instead use existing `ReadLineAsSecureString()` method which would already be implemented by different hosts and convert to a string if `-MaskInput` is used.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/13011

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed] https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5021(https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs [hunk 1]
```diff
@@ -61,29 +61,6 @@ public abstract class PSHostUserInterface
...
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string)"/>
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string, System.Management.Automation.PSCredentialTypes, System.Management.Automation.PSCredentialUIOptions)"/>
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForChoice"/>
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.Prompt"/>
-        public virtual string ReadLineMaskedAsString()
-        {
-            // Default implementation of the function to maintain backwards compatibility of the base class.
-            throw new PSNotImplementedException();
-        }
-
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:86]: This seems like a fairly large breaking change that would probably break most other hosts out there. We need think more about how to manage this, but it should probably at least be redirected in some way

**Review by rjmholt — COMMENTED** [review]
The host change makes me think we should have more SDK tests, particularly a test host implementation, so we have a good litmus test for breaking changes in the API surface

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:86]: This deleted code was added in PS7.1 preview.  I'm removing it as to not affect existing hosts.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs [hunk 2]
```diff
@@ -149,27 +149,37 @@ protected override void BeginProcessing()
                 {
                     foreach (PSObject o in result.Values)
                     {
-                        WriteObject(o);
+                        if (MaskInput && o.BaseObject is System.Security.SecureString)
+                        {
+                            WriteObject(Utils.GetStringFromSecureString((System.Security.SecureString)o.BaseObject));
+                        }
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs]: Looks like that namespace is already being used so removing all unnecessary fully qualified usage

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs (hunk 1 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:86]: Ok, that makes sense -- happy with that then

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs [hunk 3]
```diff
@@ -149,27 +149,37 @@ protected override void BeginProcessing()
                 {
                     foreach (PSObject o in result.Values)
                     {
-                        WriteObject(o);
+                        if (MaskInput && o.BaseObject is SecureString)
+                        {
+                            WriteObject(Utils.GetStringFromSecureString((SecureString)o.BaseObject));
+                        }
```
Comment (rjmholt) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs:155]: Can still eliminate an unnecessary cast:

```suggestion
                        if (MaskInput && o.BaseObject is SecureString maskedInput)
                        {
                            WriteObject(Utils.GetStringFromSecureString(maskedInput));
                        }
```

**Review by rjmholt — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry macos

**Comment by PoshChan** [comment]
@SteveL-MSFT, did not find matching build context: `PowerShell-CI-macOS`; allowed contexts: PowerShell-CI-SSH

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs [hunk 4]
```diff
@@ -149,27 +149,37 @@ protected override void BeginProcessing()
                 {
                     foreach (PSObject o in result.Values)
                     {
-                        WriteObject(o);
+                        if (MaskInput && o.BaseObject is SecureString secureString)
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs:152]:
```suggestion
                        if (MaskInput && o?.BaseObject is SecureString secureString)
```

Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs:152]: This is to prevent a potential null reference exception.

**Review by daxian-dbw — COMMENTED** [review]
LGTM except for one minor comment.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs (hunk 4 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ReadConsoleCmdlet.cs:152]: Actually, let me address that minor change, as a maintainer :)

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Can you please resolve the conflict? mac CI failing tests are marked as pending for now, so CI should pass after you rebase.

**Comment by daxian-dbw** [comment]
@PoshChan Please remind me in 1 hour

**Comment by PoshChan** [comment]
@daxian-dbw, this is the reminder you requested 1 hour ago

**Pull request merged**